### ISSUE TITLE
4.1.(0, 1) cluster does not form in the presence of neo4j-streams-4.0.2.jar

### ISSUE THREAD

**Issue by moxious** [neo4j-contrib/neo4j-streams#342]
## Actual Behavior (Mandatory)

See internal slack discussion involving dvoutila.   I independently reproduced this.  

## How to Reproduce the Problem

- Launch 4.1.0 cluster or 4.1.1
- Add neo4j-streams-4.0.2.jar.  
- With or without kafka/streams configuration in neo4j.conf, the same behavior will be seen, I did without any config (default neo4j.conf)
- Start the cluster, and then run a CREATE DATABASE command.
- Database is not created, and you will see lots of messages in the log like this:

```
2020-08-12 20:43:58.940+0000 INFO [c.n.c.c.s.CoreSnapshotService] [system] Waiting for another raft group member to publish a core state snapshot
```

## Hypothesis

Raft group for databases never properly form.   I don't see any evidence in the logs of anything bad that streams is doing, but this *related ticket* is a hint:  https://github.com/neo4j-contrib/neo4j-streams/issues/341 it appears that streams might be trying to subscribe to kafka topics or otherwise doing something with the kafka client, even in the absence of good configuration.

We've seen some issues lately where the database & cluster lifecycle is not appropriately respected by the internal availability guard.  That is to say, the DBMS may signal to plugins that a database is ready before the database is actually ready, causing streams to do something before the DBMS is ready for it.   So this is my hunch on where to look.   In particular, we need to look for code paths of how neo4j-streams becomes aware of new databases being created.  What actions does it take, and critically *when* in response to that?

## More Technical Information

(It may mention another database or the one you created, not necesarily the system database)

Behavior I'm seeing in a local cluster is that the raft group never properly forms.

One cluster member says this:

```
2020-08-12 21:49:42.358+0000 INFO [c.n.c.c.s.s.DurableStateStorage] [newdb] session-tracker state restored, up to ordinal -1
2020-08-12 21:49:42.364+0000 INFO [c.n.c.c.s.s.DurableStateStorage] [newdb] last-flushed state restored, up to ordinal -1
2020-08-12 21:49:42.367+0000 INFO [c.n.c.c.c.l.s.SegmentedRaftLog] [newdb] log started with recovered state State{prevIndex=-1, prevTerm=-1, appendIndex=-1}
2020-08-12 21:49:42.367+0000 INFO [c.n.c.c.c.m.RaftMembershipManager] [newdb] Membership state before recovery: RaftMembershipState{committed=null, appended=null, ordinal=-1}
2020-08-12 21:49:42.367+0000 INFO [c.n.c.c.c.m.RaftMembershipManager] [newdb] Recovering from: -1 to: -1
2020-08-12 21:49:42.367+0000 INFO [c.n.c.c.c.m.RaftMembershipManager] [newdb] Membership state after recovery: RaftMembershipState{committed=null, appended=null, ordinal=-1}
2020-08-12 21:49:42.368+0000 INFO [c.n.c.c.c.m.RaftMembershipManager] [newdb] Target membership: []
2020-08-12 21:49:42.368+0000 WARN [c.n.c.c.c.m.RaftMembershipManager] [newdb] Target membership [] does not contain a majority of existing raft members []. It is likely an operator removed too many core members too quickly. If not, this issue should be transient.
2020-08-12 21:49:42.388+0000 INFO [c.n.c.d.RaftMonitor] Trying bootstrap using initial members [MemberId{e10a7715}, MemberId{b1275961}, MemberId{6764dade}] and store ID StoreId{creationTime=1597268981712, randomId=-5020605660939756307, storeVersion=3471765337752883975, upgradeTime=1597268981712, upgradeTxId=-5020605660939756307}
2020-08-12 21:49:42.398+0000 INFO [c.n.c.c.s.RaftBootstrapper] [newdb] Bootstrapping database newdb for members [MemberId{e10a7715}, MemberId{b1275961}, MemberId{6764dade}]
2020-08-12 21:49:42.399+0000 INFO [c.n.c.c.s.RaftBootstrapper] [newdb] Initializing the store for database newdb using a temporary database in com.neo4j.causalclustering.core.TempBootstrapDir$Resource@33dcf233
2020-08-12 21:49:42.752+0000 INFO [c.n.c.d.a.GlobalTopologyState] Core topology for database DatabaseId{087b170c} is now: [MemberId{b1275961}, MemberId{6764dade}, MemberId{e10a7715}]
No members where lost
New members: 
  MemberId{6764dade}=CoreServerInfo{raftServer=10.142.0.105:7000, catchupServer=10.142.0.105:6000, connectorAddresses=bolt://35.243.139.186:7687,http://35.243.139.186:7474,https://35.243.139.186:7473, groups=[], startedDatabaseIds=[DatabaseId{cfb8387e}, DatabaseId{0f9b21e7}, DatabaseId{087b170c}, DatabaseId{00000000}], refuseToBeLeader=false}
  MemberId{e10a7715}=CoreServerInfo{raftServer=10.142.0.16:7000, catchupServer=10.142.0.16:6000, connectorAddresses=bolt://34.75.36.233:7687,http://34.75.36.233:7474,https://34.75.36.233:7473, groups=[], startedDatabaseIds=[DatabaseId{0f9b21e7}, DatabaseId{cfb8387e}, DatabaseId{087b170c}, DatabaseId{00000000}], refuseToBeLeader=false}
  MemberId{b1275961}=CoreServerInfo{raftServer=10.142.0.104:7000, catchupServer=10.142.0.104:6000, connectorAddresses=bolt://35.237.123.125:7687,http://35.237.123.125:7474,https://35.237.123.125:7473, groups=[], startedDatabaseIds=[DatabaseId{cfb8387e}, DatabaseId{0f9b21e7}, DatabaseId{087b170c}, DatabaseId{00000000}], refuseToBeLeader=false}
2020-08-12 21:49:42.753+0000 INFO [c.n.c.c.c.m.RaftMembershipManager] [newdb] Target membership: [MemberId{b1275961}, MemberId{6764dade}, MemberId{e10a7715}]
2020-08-12 21:49:42.753+0000 WARN [c.n.c.c.c.m.RaftMembershipManager] [newdb] Target membership [MemberId{e10a7715}, MemberId{b1275961}, MemberId{6764dade}] does not contain a majority of existing raft members []. It is likely an operator removed too many core members too quickly. If not, this issue should be transient.
2020-08-12 21:49:42.754+0000 WARN [c.n.c.c.c.m.RaftMembershipManager] [newdb] Target membership [MemberId{e10a7715}, MemberId{b1275961}, MemberId{6764dade}] does not contain a majority of existing raft members []. It is likely an operator removed too many core members too quickly. If not, this issue should be transient.
2020-08-12 21:51:07.807+0000 INFO [c.n.c.c.c.l.s.Segments] [system] Pruning SegmentFile{file=raft.log.0, header=SegmentHeader{formatVersion=2, recordOffset=56, prevFileLastIndex=-1, segmentNumber=0, prevIndex=-1, prevTerm=-1}}
2020-08-12 21:51:24.008+0000 INFO [c.n.c.c.c.l.s.Segments] [neo4j] Pruning SegmentFile{file=raft.log.0, header=SegmentHeader{formatVersion=2, recordOffset=56, prevFileLastIndex=-1, segmentNumber=0, prevIndex=-1, prevTerm=-1}}
2020-08-12 21:51:24.021+0000 INFO [c.n.c.c.c.l.s.Segments] [foo] Pruning SegmentFile{file=raft.log.0, header=SegmentHeader{formatVersion=2, recordOffset=56, prevFileLastIndex=-1, segmentNumber=0, prevIndex=-1, prevTerm=-1}}
```

the others say this:

```
2020-08-12 21:49:42.421+0000 INFO [c.n.c.d.RaftMonitor] Trying bootstrap using initial members [MemberId{e10a7715}, MemberId{b1275961}, MemberId{6764dade}] and store ID StoreId{creationTime=1597268981712, randomId=-5020605660939756307, storeVersion=3471765337752883975, upgradeTime=1597268981712, upgradeTxId=-5020605660939756307}
2020-08-12 21:49:42.430+0000 INFO [c.n.c.c.s.CoreSnapshotService] [newdb] Waiting for another raft group member to publish a core state snapshot
2020-08-12 21:49:42.824+0000 INFO [c.n.c.d.a.GlobalTopologyState] Core topology for database DatabaseId{087b170c} is now: [MemberId{b1275961}, MemberId{6764dade}, MemberId{e10a7715}]
No members where lost
New members: 
  MemberId{6764dade}=CoreServerInfo{raftServer=10.142.0.105:7000, catchupServer=10.142.0.105:6000, connectorAddresses=bolt://35.243.139.186:7687,http://35.243.139.186:7474,https://35.243.139.186:7473, groups=[], startedDatabaseIds=[DatabaseId{087b170c}, DatabaseId{00000000}, DatabaseId{0f9b21e7}, DatabaseId{cfb8387e}], refuseToBeLeader=false}
  MemberId{e10a7715}=CoreServerInfo{raftServer=10.142.0.16:7000, catchupServer=10.142.0.16:6000, connectorAddresses=bolt://34.75.36.233:7687,http://34.75.36.233:7474,https://34.75.36.233:7473, groups=[], startedDatabaseIds=[DatabaseId{cfb8387e}, DatabaseId{0f9b21e7}, DatabaseId{087b170c}, DatabaseId{00000000}], refuseToBeLeader=false}
  MemberId{b1275961}=CoreServerInfo{raftServer=10.142.0.104:7000, catchupServer=10.142.0.104:6000, connectorAddresses=bolt://35.237.123.125:7687,http://35.237.123.125:7474,https://35.237.123.125:7473, groups=[], startedDatabaseIds=[DatabaseId{cfb8387e}, DatabaseId{0f9b21e7}, DatabaseId{087b170c}, DatabaseId{00000000}], refuseToBeLeader=false}
2020-08-12 21:49:42.825+0000 INFO [c.n.c.c.c.m.RaftMembershipManager] [newdb] Target membership: [MemberId{b1275961}, MemberId{6764dade}, MemberId{e10a7715}]
2020-08-12 21:49:42.825+0000 WARN [c.n.c.c.c.m.RaftMembershipManager] [newdb] Target membership [MemberId{e10a7715}, MemberId{b1275961}, MemberId{6764dade}] does not contain a majority of existing raft members []. It is likely an operator removed too many core members too quickly. If not, this issue should be transient.
2020-08-12 21:49:42.826+0000 WARN [c.n.c.c.c.m.RaftMembershipManager] [newdb] Target membership [MemberId{e10a7715}, MemberId{b1275961}, MemberId{6764dade}] does not contain a majority of existing raft members []. It is likely an operator removed too many core members too quickly. If not, this issue should be transient.
2020-08-12 21:49:52.435+0000 INFO [c.n.c.c.s.CoreSnapshotService] [newdb] Waiting for another raft group member to publish a core state snapshot
2020-08-12 21:50:02.439+0000 INFO [c.n.c.c.s.CoreSnapshotService] [newdb] Waiting for another raft group member to publish a core state snapshot
2020-08-12 21:50:12.442+0000 INFO [c.n.c.c.s.CoreSnapshotService] [newdb] Waiting for another raft group member to publish a core state snapshot
2020-08-12 21:50:22.446+0000 INFO [c.n.c.c.s.CoreSnapshotService] [newdb] Waiting for another raft group member to publish a core state snapshot
2020-08-12 21:50:32.450+0000 INFO [c.n.c.c.s.CoreSnapshotService] [newdb] Waiting for another raft group member to publish a core state snapshot
```

**Comment by voutilad** [comment]
Alternative way to reproduce is to install the plugin prior to first start of the cluster. The cluster will never form and the `CoreSnapshotService` will log the same "Waiting for another raft group member..." message for the `system` database.

**Comment by moxious** [comment]
Another thing to check is the release notes for the 4.1 series.  Note that new features went in about database leader "balancing" https://neo4j.com/release-notes/neo4j-4-1-0/

It's possible that under 4.0 there was one machine always the leader for most/all databases, and 4.1 I think is intentionally "spreading out the leader" role to help maximize write throughput; but when we need to monitor the availability of a lot of databases, and the leaders are spread out, this has implications for where streams code is executing

**Comment by tanvp112** [comment]
Doesn't matter if it is a cluster or not, here's a quick way to reproduce this issue, just follow the example by updating the image to 4.1.1-enterprise: https://neo4j.com/docs/labs/neo4j-streams/current/docker/

![image](https://user-images.githubusercontent.com/57385114/91296544-5fd0fe00-e7cf-11ea-89f7-ee2b1d90b3b6.png)

**Comment by voutilad** [comment]
Still a problem with the latest plugin version. Found this stack trace in the debug logs showing it seems related to trying to inspect the databases before the database fully exists (I snipped out a LARGE swath of steps to help get to the usage of the streams library):

```
2020-09-02 14:24:23.466+0000 DEBUG [s.c.StreamsConfig] Neo4j Streams configuration reloaded from neo4j.conf file: {}
2020-09-02 14:24:25.461+0000 DEBUG [o.n.c.i.CommunityCompilerFactory] [system] Index not found for indexPropertyExistsSelectivity No index was found
org.neo4j.internal.kernel.api.exceptions.schema.IndexNotFoundKernelException: No index was found
	at org.neo4j.kernel.impl.newapi.AllStoreHolder.assertValidIndex(AllStoreHolder.java:1033)
	at org.neo4j.kernel.impl.newapi.AllStoreHolder.indexSize(AllStoreHolder.java:713)
	at org.neo4j.cypher.internal.spi.TransactionBoundGraphStatistics$BaseTransactionBoundGraphStatistics.indexPropertyExistsSelectivity(TransactionBoundGraphStatistics.scala:89)
	at org.neo4j.cypher.internal.planner.spi.DelegatingGraphStatistics.indexPropertyExistsSelectivity(GraphStatistics.scala:80)
	at org.neo4j.cypher.internal.planner.spi.InstrumentedGraphStatistics.indexPropertyExistsSelectivity(InstrumentedGraphStatistics.scala:98)
<SNIP SNIP SNIP>
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.execute(TransactionImpl.java:236)
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.execute(TransactionImpl.java:225)
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.execute(TransactionImpl.java:219)
	at streams.extensions.DatabaseManagementServiceExtensionsKt.getDefaultDbName(DatabaseManagementServiceExtensions.kt:12)
	at streams.config.StreamsConfig.defaultDbName(StreamsConfig.kt:123)
	at streams.config.StreamsConfig.isDefaultDb(StreamsConfig.kt:125)
	at streams.StreamsEventRouterConfiguration$Companion.from(StreamsEventRouterConfiguration.kt:54)
	at streams.StreamsEventRouterLifecycle.start(StreamsExtensionFactory.kt:54)
	at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:442)
	at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:110)
	at org.neo4j.kernel.extension.AbstractExtensions.start(AbstractExtensions.java:82)
	at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:442)
	at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:110)
	at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:442)
	at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:110)
	at org.neo4j.kernel.database.Database.start(Database.java:481)
	at org.neo4j.dbms.database.AbstractDatabaseManager.startDatabase(AbstractDatabaseManager.java:187)
	at com.neo4j.dbms.database.MultiDatabaseManager.forSingleDatabase(MultiDatabaseManager.java:134)
	at com.neo4j.dbms.database.MultiDatabaseManager.startDatabase(MultiDatabaseManager.java:119)
	at com.neo4j.dbms.Transition$Prepared.doTransitionAction(Transition.java:101)
	at com.neo4j.dbms.Transition$Prepared.doTransition(Transition.java:88)
	at com.neo4j.dbms.DbmsReconciler.doTransitionStep(DbmsReconciler.java:346)
	at com.neo4j.dbms.DbmsReconciler.doTransitionStep(DbmsReconciler.java:347)
	at com.neo4j.dbms.DbmsReconciler.doTransitionStep(DbmsReconciler.java:347)
	at com.neo4j.dbms.DbmsReconciler.lambda$doTransitions$11(DbmsReconciler.java:315)
	at com.neo4j.dbms.DbmsReconciler.namedJob(DbmsReconciler.java:326)
	at com.neo4j.dbms.DbmsReconciler.doTransitions(DbmsReconciler.java:316)
	at com.neo4j.dbms.DbmsReconciler.lambda$doTransitions$9(DbmsReconciler.java:307)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by conker84** [comment]
Hi guys thanks a lot for reporting this can you try with the following jar?
[neo4j-streams-4.0.2.jar.zip](https://github.com/neo4j-contrib/neo4j-streams/files/5167089/neo4j-streams-4.0.2.jar.zip)

**Comment by conker84** [comment]
You can tune this parameter in order to startup your instance
`streams.systemdb.wait.timeout=<ms that the plugin waits the systemdb to be ready, default=10000>`

**Comment by voutilad** [comment]
Using that jar, Neo4j 4.1 properly cold starts now and (at least in a single-instance deployment) I can create a new database.

One nitpick in the logs...mixing English variations of `s/z`:
```
2020-09-03 12:32:38.514+0000 INFO  [junk] Initialising the Streams Source module
2020-09-03 12:32:38.531+0000 INFO  [junk] Initializing Kafka Connector
```

I haven't tested a causal cluster deployment, but if you get the changes into the 4.0 branch I'll build a copy and try it later this week.

**Comment by tanvp112** [comment]
Tested with neo4j-streams-4.0.3.jar using the same example above: Neo4j server couldn't start-up on IPv4 interface.

![image](https://user-images.githubusercontent.com/57385114/92915878-91002e00-f45f-11ea-908e-a020aa0ab53d.png)

![image](https://user-images.githubusercontent.com/57385114/92916308-f7854c00-f45f-11ea-9a98-6f41eeeb8e1c.png)

**Comment by tanvp112** [comment]
> You can tune this parameter in order to startup your instance
> `streams.systemdb.wait.timeout=<ms that the plugin waits the systemdb to be ready, default=10000>`

Hi Conker84,

The docker examples (tested up to source/sink) can run as expected with your jar. I deployed to docker swarm and it works as expected as well. Hope to see your fix goes into the mainline soon.

Thanks.

**Comment by tanvp112** [comment]
> Hi guys thanks a lot for reporting this can you try with the following jar?
> [neo4j-streams-4.0.2.jar.zip](https://github.com/neo4j-contrib/neo4j-streams/files/5167089/neo4j-streams-4.0.2.jar.zip)

Hi Conker84,

The docker examples (tested up to source/sink) can run as expected with your jar. I deployed to docker swarm and it works as expected as well. Hope to see your fix goes into the mainline soon.

Thanks.

**Comment by conker84** [comment]
closed via 614a7ab
